src/relay/backend/aot_executor_codegen.cc

```diff
@@ -637,6 +681,7 @@ class AOTExecutorCodegen : public MixedModeVisitor {
       main_signature_.push_back(tir::Var("output", DataType::Handle()));
     }
 
+    CollectDeviceVariables(lowered_mod->GetAttr<Map<GlobalVar, String>>("device_contexts").value());
```

Comment:
I think the blocker for pulling the device (or, after #9326 the SEScope) for the callee is just that we're not carrying those annotations over into the PrimFuncs, right? Would you consider trying that after #9326? I've unfortunately hit my limit for any more hackery within the te_compiler.cc.

Then, afaikt  you're doing what I've also had to do in the VM compiller.cc, which is allocate globally unique tags (just ints in the VM) for every unique SEScope, and build a global map from those tags to their actual devices (just an array of Devices in the VM). Everything you've done to achieve that part looks perfectly sensible to me.

Reply:
Sorry @mbs-octoml, I feel like I'm justified in just one more hack for this :crying_cat_face: 

Unfortunately we also need to rectify the `kCompiler` vs `Target` mix at the same time as this current implementation uses what I'm calling a hybrid `Target` (a registered `Target` which is still using most of the BYOC infrastructure). The eventual aim is to get rid of `kCompiler` and move fully to `Target` so that the above annotations end up on all `Function`s of relevance and then get rid of this. 

Internally to the TECompiler, the representation of the BYOC function gets changed to a `GlobalVar` after the name gets uniquified, and the annotation gets removed, so until everything gets persisted naturally all the way down I can't get to it sensibly without this :crying_cat_face: 

Or I could re-use some way of getting to the TE Compiler internal cache if you'd like? :smiling_imp:

Reply:
> Internally to the TECompiler, the representation of the BYOC function gets changed to a GlobalVar after the name gets uniquified, and the annotation gets removed, so until everything gets persisted naturally all the way down I can't get to it sensibly without this

Hmm, ok, so you're still on the BYOC compilation notion of 'target' while I'm on the, well, Target notion of 'target'.
Even so, you can avoid the IRModule attribute map by annotating the PrimFunc, right? That's at least closer to the ultimate goal.

Reply:
The Call gets translated to `Call(GlobalVar)`, the `PrimFunc` is just a stub one for the TE Compiler and because the function name may change I can't find the eventual `Function` again. Even if I could guarantee the `GlobalVar` didn't change, and find the `Function`, the annotation gets stripped by `LowerExternalFunctions`. When this gets revisited the aim would be to not remove the annotation, so it can be found again later, I believe it's done now because `kCompiler` causes different paths in different parts of the compiler which are undesirable.

Reply:
Ack. Thanks for the extra comment.